src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java

```diff
@@ -0,0 +1,203 @@
+    return format;
+  }
+
+  public String getTableName() {
+    return options.getTableName();
+  }
+
+  /** @return List of column names specified in COPY statement, if provided. */
+  public List<String> getCopyColumnNames() {
+    return options.getColumnNames();
+  }
+
+  /** @return Format type specified in COPY statement, if provided. */
+  public String getFormatType() {
+    return options.getFormat().toString();
+  }
+
+  /** @return Null string specified in COPY statement, if provided. */
+  public String getNullString() {
+    return options.getNullString();
+  }
+
+  /** @return Delimiter character specified in COPY statement, if provided. */
+  public char getDelimiterChar() {
+    return options.getDelimiter();
+  }
+
+  /** @return Escape character specified in COPY statement, if provided. */
+  public char getEscapeChar() {
+    return options.getEscape();
+  }
+
+  /** @return Quote character specified in COPY statement, if provided. */
+  public char getQuoteChar() {
+    return options.getQuote();
+  }
+
+  /** @return True if copy data contains a header, false otherwise. */
+  public boolean hasHeader() {
+    return options.getHeader();
+  }
+
+  public MutationWriter getMutationWriter() {
+    return this.mutationWriter;
+  }
+
+  private void verifyCopyColumns() throws SQLException {
+    if (options.getColumnNames().size() > this.tableColumns.size()) {
+      throw new SQLException("Number of copy columns provided exceed table column count");
+    }
+    LinkedHashMap<String, TypeCode> tempTableColumns = new LinkedHashMap<>();
+    // Verify that every copy column given is a valid table column name
+    for (String copyColumn : options.getColumnNames()) {
+      if (!this.tableColumns.containsKey(copyColumn)) {
+        throw new SQLException(
+            "Column \"" + copyColumn + "\" of relation \"" + getTableName() + "\" does not exist");
+      }
+      // Update table column ordering with the copy column ordering
+      tempTableColumns.put(copyColumn, tableColumns.get(copyColumn));
+    }
+    this.tableColumns = tempTableColumns;
+  }
+
+  private static TypeCode parseSpannerDataType(String columnType) {
+    // Eliminate size modifiers in column type (e.g. STRING(100), BYTES(50), etc.)
+    int index = columnType.indexOf("(");
+    columnType = (index > 0) ? columnType.substring(0, index) : columnType;
+    TypeCode type = TypeCode.valueOf(columnType);
+    if (type == null) {
+      throw new IllegalArgumentException(
+          "Unrecognized or unsupported column data type: " + columnType);
+    }
+    return type;
+  }
+
+  private void queryInformationSchema() throws SQLException {
+    Map<String, TypeCode> tableColumns = new LinkedHashMap<>();
+    PreparedStatement statement =
+        this.connection.prepareStatement(
+            "SELECT "
+                + COLUMN_NAME
+                + ", "
+                + SPANNER_TYPE
+                + " FROM information_schema.columns WHERE table_name = ?");
```

Comment:
does the table name here preserve quotes? If not, can you add quotes in the INFORMATION_SCHEMA query? Then it should work for "foo" and "FOO"

Reply:
Done.